drivers/dai/nxp/sai.c

```diff
@@ -0,0 +1,758 @@
+	sai_tx_rx_sw_enable_disable(dir, data, false);
+
+	return 0;
+}
+
+static int sai_trigger_stop(const struct device *dev,
+			    enum dai_dir dir)
+{
+	struct sai_data *data;
+	int ret;
+	uint32_t old_state;
+
+	data = dev->data;
+	old_state = sai_get_state(dir, data);
+
+	if (dir != DAI_DIR_RX && dir != DAI_DIR_TX) {
+		LOG_ERR("invalid direction: %d", dir);
+		return -EINVAL;
+	}
+
+	/* attempt to change state */
+	ret = sai_update_state(dir, data, DAI_STATE_STOPPING);
+	if (ret < 0) {
+		LOG_ERR("failed to transition to STOPPING from %d. Reason: %d",
+			sai_get_state(dir, data), ret);
+		return ret;
+	}
+
+	LOG_DBG("stop on direction %d", dir);
+
+	if (old_state == DAI_STATE_PAUSED) {
+		/* if SAI was previously paused then all that's
+		 * left to do is disable the DMA requests and
+		 * the data line.
+		 */
+		goto out_dline_disable;
+	}
+
+	ret = sai_tx_rx_disable(data, dir);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* update the software state of TX/RX */
+	sai_tx_rx_sw_enable_disable(dir, data, false);
+
+out_dline_disable:
+	/* disable TX/RX data line */
+	sai_tx_rx_set_dline_mask(dir, data->regmap, 0x0);
+
+	/* disable DMA requests */
+	SAI_TX_RX_DMA_ENABLE_DISABLE(dir, data->regmap, false);
+
+	/* disable error interrupt */
+	SAI_TX_RX_ENABLE_DISABLE_IRQ(dir, data->regmap,
+				     kSAI_FIFOErrorInterruptEnable, false);
+
+	return 0;
+}
+
+static int sai_trigger_start(const struct device *dev,
+			     enum dai_dir dir)
+{
+	struct sai_data *data;
+	uint32_t old_state;
+	int ret;
+
+	data = dev->data;
+	old_state = sai_get_state(dir, data);
+
+	/* TX and RX should be triggered independently */
+	if (dir != DAI_DIR_RX && dir != DAI_DIR_TX) {
+		LOG_ERR("invalid direction: %d", dir);
+		return -EINVAL;
+	}
+
+	/* attempt to change state */
+	ret = sai_update_state(dir, data, DAI_STATE_RUNNING);
+	if (ret < 0) {
+		LOG_ERR("failed to transition to RUNNING from %d. Reason: %d",
+			sai_get_state(dir, data), ret);
+		return ret;
+	}
+
+	if (old_state == DAI_STATE_PAUSED) {
+		/* if the SAI has been paused then there's no
+		 * point in issuing a software reset. As such,
+		 * skip this part and go directly to the TX/RX
+		 * enablement.
+		 */
+		goto out_enable_tx_rx;
+	}
+
+	LOG_DBG("start on direction %d", dir);
+
+	if (dir == DAI_DIR_RX) {
+		/* this is fine because TX is async so it won't be
+		 * affected by an RX software reset.
+		 */
+		SAI_TX_RX_SW_RESET(dir, data->regmap);
+
+		/* do a TX software reset only if not already enabled */
+		if (!data->tx_enabled) {
+			SAI_TX_RX_SW_RESET(DAI_DIR_TX, data->regmap);
+		}
+	} else {
+		/* a software reset should be issued for TX
+		 * only if RX was not already enabled.
+		 */
+		if (!data->rx_enabled) {
+			SAI_TX_RX_SW_RESET(dir, data->regmap);
+		}
+	}
+
+	/* enable error interrupt */
+	SAI_TX_RX_ENABLE_DISABLE_IRQ(dir, data->regmap,
+				     kSAI_FIFOErrorInterruptEnable, true);
+
+	/* TODO: is there a need to write some words to the FIFO to avoid starvation? */
```

Comment:
This comment was left from the time I was testing the driver (using DMA domain) (although the domain we use shouldn't make a difference as the underrun would probably happen during start() if there was a problem). For now, I haven't managed to get a FIFO underrun because of this but yeah I agree with you on the testing bit.